editor/src/components/canvas/canvas-strategies/interaction-state.ts

```diff
@@ -0,0 +1,321 @@
+    startingMetadata: metadata ?? {},
+  }
+}
+
+export function createInteractionViaMouse(
+  mouseDownPoint: CanvasPoint,
+  modifiers: Modifiers,
+  activeControl: CanvasControlType,
+): InteractionSessionWithoutMetadata {
+  return {
+    interactionData: {
+      type: 'DRAG',
+      dragStart: mouseDownPoint,
+      drag: null,
+      prevDrag: null,
+      dragThresholdPassed: false,
+      originalDragStart: mouseDownPoint,
+      modifiers: modifiers,
+    },
+    activeControl: activeControl,
+    sourceOfUpdate: activeControl,
+    lastInteractionTime: Date.now(),
+    userPreferredStrategy: null,
+    startedAt: Date.now(),
+    globalTime: Date.now(),
+  }
+}
+
+function dragExceededThreshold(drag: CanvasVector): boolean {
+  const xDiff = Math.abs(drag.x)
+  const yDiff = Math.abs(drag.y)
+  return xDiff > MoveIntoDragThreshold || yDiff > MoveIntoDragThreshold
+}
+
+export function updateInteractionViaMouse(
+  currentState: InteractionSession,
+  drag: CanvasVector,
+  modifiers: Modifiers,
+  sourceOfUpdate: CanvasControlType | null, // If null it means the active control is the source
+): InteractionSessionWithoutMetadata {
+  if (currentState.interactionData.type === 'DRAG') {
+    const dragThresholdPassed =
+      currentState.interactionData.dragThresholdPassed || dragExceededThreshold(drag)
+    return {
+      interactionData: {
+        type: 'DRAG',
+        dragStart: currentState.interactionData.dragStart,
+        drag: dragThresholdPassed ? drag : null,
+        prevDrag: currentState.interactionData.drag,
+        dragThresholdPassed: dragThresholdPassed,
+        originalDragStart: currentState.interactionData.originalDragStart,
+        modifiers: modifiers,
+      },
+      activeControl: currentState.activeControl,
+      sourceOfUpdate: sourceOfUpdate ?? currentState.activeControl,
+      lastInteractionTime: Date.now(),
+      userPreferredStrategy: currentState.userPreferredStrategy,
+      startedAt: currentState.startedAt,
+      globalTime: Date.now(),
+    }
+  } else {
+    return {
+      ...currentState,
+      globalTime: Date.now(),
+    }
+  }
+}
+
+export function createInteractionViaKeyboard(
+  keysPressed: Array<KeyCharacter>,
+  modifiers: Modifiers,
+  activeControl: CanvasControlType,
+): InteractionSessionWithoutMetadata {
+  return {
+    interactionData: {
+      type: 'KEYBOARD',
+      keysPressed: keysPressed,
+      modifiers: modifiers,
+    },
+    activeControl: activeControl,
+    sourceOfUpdate: activeControl,
+    lastInteractionTime: Date.now(),
+    userPreferredStrategy: null,
+    startedAt: Date.now(),
+    globalTime: Date.now(),
+  }
+}
+
+export function updateInteractionViaKeyboard(
+  currentState: InteractionSession,
+  addedKeysPressed: Array<KeyCharacter>,
+  keysReleased: Array<KeyCharacter>,
+  modifiers: Modifiers,
+  sourceOfUpdate: CanvasControlType,
+): InteractionSessionWithoutMetadata {
+  if (currentState.interactionData.type === 'KEYBOARD') {
```

Comment:
Can we replace this with a `switch` statement please so that we can ensure completeness with the usual:
```ts
    default:
      const _exhaustiveCheck: never = currentState.interactionData
      throw new Error(`Unhandled interactionData type ${JSON.stringify(currentState.interactionData)}`)
```